# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.cpp
# Commit: 9111720170a2
# Full Hash: 9111720170a283e7c3801856d5c113a2f17caf20
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2022-05-06 09:50:52
# Regressor Bug: 1642412
# File Overlap Count: 7
# Description:
#   Bug 1762171 - wasm: Fix cranelift build for table optimizations. r=yury
#   
#   This commit adds back the builtins that were removed in
#   bug 1642412.
#   
# ==============================================================================

diff -r 2340f31f2559 -r 9111720170a2 js/src/wasm/WasmInstance.cpp
--- a/js/src/wasm/WasmInstance.cpp	Thu May 05 21:49:46 2022 +0000
+++ b/js/src/wasm/WasmInstance.cpp	Thu May 05 21:49:46 2022 +0000
@@ -996,9 +996,9 @@
   return 0;
 }
 
-/* static */ void* Instance::tableGetFunc(Instance* instance, uint32_t index,
-                                          uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableGetFunc.failureMode == FailureMode::FailOnInvalidRef);
+/* static */ void* Instance::tableGet(Instance* instance, uint32_t index,
+                                      uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableGet.failureMode == FailureMode::FailOnInvalidRef);
 
   JSContext* cx = instance->cx();
   const Table& table = *instance->tables()[tableIndex];
@@ -1007,13 +1007,19 @@
     return AnyRef::invalid().forCompiledCode();
   }
 
-  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
-  MOZ_RELEASE_ASSERT(!table.isAsmJS());
-  RootedFunction fun(cx);
-  if (!table.getFuncRef(cx, index, &fun)) {
-    return AnyRef::invalid().forCompiledCode();
+  switch (table.repr()) {
+    case TableRepr::Ref:
+      return table.getAnyRef(index).forCompiledCode();
+    case TableRepr::Func: {
+      MOZ_RELEASE_ASSERT(!table.isAsmJS());
+      RootedFunction fun(cx);
+      if (!table.getFuncRef(cx, index, &fun)) {
+        return AnyRef::invalid().forCompiledCode();
+      }
+      return FuncRef::fromJSFunction(fun).forCompiledCode();
+    }
   }
-  return FuncRef::fromJSFunction(fun).forCompiledCode();
+  MOZ_CRASH("Should not happen");
 }
 
 /* static */ uint32_t Instance::tableGrow(Instance* instance, void* initValue,
@@ -1042,9 +1048,9 @@
   return oldSize;
 }
 
-/* static */ int32_t Instance::tableSetFunc(Instance* instance, uint32_t index,
-                                            void* value, uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableSetFunc.failureMode == FailureMode::FailOnNegI32);
+/* static */ int32_t Instance::tableSet(Instance* instance, uint32_t index,
+                                        void* value, uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableSet.failureMode == FailureMode::FailOnNegI32);
 
   JSContext* cx = instance->cx();
   Table& table = *instance->tables()[tableIndex];
@@ -1054,12 +1060,26 @@
     return -1;
   }
 
-  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
-  MOZ_RELEASE_ASSERT(!table.isAsmJS());
-  table.fillFuncRef(index, 1, FuncRef::fromCompiledCode(value), cx);
+  switch (table.repr()) {
+    case TableRepr::Ref:
+      table.fillAnyRef(index, 1, AnyRef::fromCompiledCode(value));
+      break;
+    case TableRepr::Func:
+      MOZ_RELEASE_ASSERT(!table.isAsmJS());
+      table.fillFuncRef(index, 1, FuncRef::fromCompiledCode(value), cx);
+      break;
+  }
+
   return 0;
 }
 
+/* static */ uint32_t Instance::tableSize(Instance* instance,
+                                          uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableSize.failureMode == FailureMode::Infallible);
+  Table& table = *instance->tables()[tableIndex];
+  return table.length();
+}
+
 /* static */ void* Instance::refFunc(Instance* instance, uint32_t funcIndex) {
   MOZ_ASSERT(SASigRefFunc.failureMode == FailureMode::FailOnInvalidRef);
   JSContext* cx = instance->cx();